ethexe/ethereum/src/router/mod.rs

```diff
@@ -117,28 +117,46 @@ impl Router {
 
     pub async fn wait_code_validation(&self, code_id: CodeId) -> Result<bool> {
         let filter = Filter::new().address(*self.instance.address());
-        let mut router_events = self
-            .instance
-            .provider()
-            .subscribe_logs(&filter)
-            .await?
-            .into_stream();
-
-        let code_id = code_id.into_bytes();
+        let provider = self.instance.provider();
 
-        while let Some(log) = router_events.next().await {
-            if let Some(signatures::CODE_GOT_VALIDATED) = log.topic0().cloned() {
-                let event = crate::decode_log::<IRouter::CodeGotValidated>(&log)?;
+        match provider.client().pubsub_frontend() {
+            Some(_) => {
+                let mut router_events = provider.subscribe_logs(&filter).await?.into_stream();
 
-                if event.codeId == code_id {
-                    return Ok(event.valid);
+                while let Some(log) = router_events.next().await {
+                    if Self::is_validated_code(log, code_id)? {
+                        return Ok(true);
+                    }
                 }
             }
-        }
+            None => loop {
+                let mut router_events = provider.watch_logs(&filter).await?.into_stream();
+
+                while let Some(logs) = router_events.next().await {
+                    for log in logs {
+                        if Self::is_validated_code(log, code_id)? {
+                            return Ok(true);
+                        }
+                    }
+                }
+            },
+        };
 
         Err(anyhow!("Failed to define if code is validated"))
     }
 
+    fn is_validated_code(log: Log, code_id: CodeId) -> Result<bool> {
```

Comment:
I think you have 3 statuses here instead of 2:
1) Some(true) - it's code id and is valid
2) Some(false) - it's code id and is not valid
3) None - it's not got validated event or for another code

Reply:
actually None case is unreachable since it checks the first topic inside